kernel/src/engine/arrow_footer_skipping.rs

```diff
@@ -0,0 +1,359 @@
+//! An implementation of parquet row group skipping using data skipping predicates.
+use crate::expressions::{BinaryOperator, Expression, Scalar, UnaryOperator, VariadicOperator};
+use crate::schema::{DataType, PrimitiveType};
+use parquet::arrow::arrow_reader::ArrowReaderBuilder;
+use parquet::file::metadata::RowGroupMetaData;
+use parquet::file::statistics::Statistics;
+use parquet::schema::types::{ColumnDescPtr, ColumnPath};
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet};
+
+pub fn filter_row_groups<T>(
+    reader: ArrowReaderBuilder<T>,
+    filter: &Expression,
+) -> ArrowReaderBuilder<T> {
+    let indices = reader
+        .metadata()
+        .row_groups()
+        .iter()
+        .enumerate()
+        .filter_map(|(index, row_group)| {
+            // We can only skip a row group if the filter is false (true/null means keep)
+            let keep = !matches!(RowGroupFilter::apply(filter, row_group), Some(false));
+            keep.then_some(index)
```

Comment:
I'm not sure those are equivalent? The intent is to keep _unless_ it's `Some(false)`. So `None` and `Some(true)` should both produce the same result. Maybe this?
```suggestion
            let keep = !RowGroupFilter::apply(filter, row_group).is_some_and(|v| !v);
            keep.then_some(index)
```

Reply:
This would be a nice place to use `Option::is_none_or`, but that's not stable rust yet:
```suggestion
            RowGroupFilter::apply(filter, row_group).is_none_or(|v| v).then_some(index)
```

Reply:
> This would be a nice place to use `Option::is_none_or`, but that's not stable rust yet:

You could use just a let chain 
```rust
if let Some(false) = RowGroupFilter::apply(filter, row_group) {
  Some(index)
} else {
  None
}
```

Reply:
This code has shifted around quite a bit. I still use `matches!(...)` macro, but the result is a plain bool now instead of `Option<bool>`:
```rust
mpl<'a> RowGroupFilter<'a> {
    /// Applies a filtering expression to a row group. Return value false means to skip it. 
    fn apply(filter: &Expression, row_group: &'a RowGroupMetaData) -> bool {
        let field_indices = compute_field_indices(row_group.schema_descr().columns(), filter);
        let result = Self {
            row_group,
            field_indices,
        }
        .apply_sql_where(filter);
        !matches!(result, Some(false))
    }
```

What do you think?